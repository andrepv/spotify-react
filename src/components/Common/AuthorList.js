import React, { Component } from "react";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";

import { ARTIST } from "constants/RouteConstants";

export default class AuthorList extends Component {
  shouldComponentUpdate(nextProps) {
    return this.props.authors !== nextProps.authors;
  }

  render() {
    const {authors, handler} = this.props;
    return authors.map((author, index) => {
      return (
        <React.Fragment key={author.id}>
          <Link to={`${ARTIST}/${author.id}`}>
            <span onClick={handler}>
              {author.name}
            </span>
          </Link>
          {index + 1 !== authors.length && ", "}
        </React.Fragment>
      );
    });
  }
}

AuthorList.propTypes = {
  authors: PropTypes.array.isRequired,
  handler: PropTypes.func,
};