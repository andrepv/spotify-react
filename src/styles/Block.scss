@import
"variables",
"modularscale";

.block {
  padding: 0 10px;
  &__media {
    margin-bottom: 12px;
  }
  &__img {
    border-radius: $border-radius;
    width: 100%;
    padding-bottom: 100%;
  }
  &__title {
    font-size: ms(-2);
    font-weight: 500;
    color: $text-dark;
    margin-bottom: 2px;
    word-break: break-word;
  }
  &__subtitle {
    font-size: ms(-3);
    color: $text-secondary;
  }
  &-header {
    display: flex;
    justify-content: space-between;
    align-items: flex-end;
    margin-bottom: 30px;
    flex-wrap: wrap;
    &__description {
      font-weight: 500;
      color: $text-secondary;
    }
    &__link {
      margin-bottom: 7px;
      svg {
        vertical-align: middle;
        line-height: 0;
        display: inline-block;
        max-height: 18px;
        width: 16px;
        margin-left: 5px;
      }
    }
  }
}

.blocks {
  &-container {
    display: flex;
    flex-wrap: wrap;
    margin: 0 -10px;
    &_cols-5 .block {
      width: calc(100% / 5);
    }
    .block {
      width: calc(100% / 4);
      margin-bottom: 40px;
      @media (max-width: 900px) {
        width: calc(100% / 3);
      }
      @media (max-width: 700px) {
        width: calc(100% / 2);
      }
      @media (max-width: 450px) {
        width: 100%;
      }
      &__img {
        box-shadow: $shadows;
      }
    }
  }
}
