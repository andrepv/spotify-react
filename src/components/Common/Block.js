import React, { Component } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import cd from "images/cd.png";
import AuthorList from "components/Common/AuthorList";
import LazyLoad from "components/Common/LazyLoad";
import "styles/Block.scss";

export default class Block extends Component {
  shouldComponentUpdate(nextProps) {
    return (
      this.props.id !== nextProps.id ||
      this.props.name !== nextProps.name ||
      this.props.image !== nextProps.image ||
      this.props.meta !== nextProps.meta
    );
  }

  render() {
    const {name = "", id, image, meta, type, handler} = this.props;
    return (
      <div className="block">
        <div className="block__media">
          <LazyLoad className="block__img">
            <div
              className={`block__img block__img_root bg-${
                !image ? "empty" : "center"
              }`}
              style={{backgroundImage: `url(${image || cd})`}}
            ></div>
          </LazyLoad>
        </div>
        <div className="block__content">
          <Link to={{
            pathname: `/${type}/${id}`,
            state: {modal: true, type},
          }}>
            <span
              className="block__title"
              onClick={handler}
            >
              {name}
            </span>
          </Link>
          <p className="block__subtitle">
            {
              type === "album"
              ? <AuthorList authors={meta} handler={handler}/>
              : meta
            }
          </p>
        </div>
      </div>
    );
  }
}

Block.propTypes = {
  image: PropTypes.string,
  name: PropTypes.string.isRequired,
  meta: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
  ]).isRequired,
  type: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  handler: PropTypes.func,
};