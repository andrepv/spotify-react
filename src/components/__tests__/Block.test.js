import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

import Block from "components/Common/Block";
import AuthorList from "components/Common/AuthorList";
import cd from "images/cd.png";

describe("<Block />", () => {
  it("renders correctly", () => {
    const props = {
      image: "test.png",
      name: "test",
      meta: "0 tracks",
      type: "playlist",
      id: "1",
      handler: jest.fn(),
    };
    const component = renderer.create(
      <Router><Block {...props}/></Router>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("check the component without an image", () => {
    const props = {
      image: "",
      name: "test",
      meta: "0 tracks",
      type: "playlist",
      id: "1",
    };
    const wrapper = shallow(<Block {...props} />);
    const target = wrapper.find(".block__img_root");

    expect(target.hasClass("bg-empty")).toEqual(true);
    expect(target.prop("style")).toHaveProperty(
      "backgroundImage", `url(${cd})`
    );
  });

  it("should display the authors as metadata", () => {
    const props = {
      image: "test.png",
      name: "test",
      meta: [{name: "John Doe", id: "1"}],
      type: "album",
      id: "1",
    };
    const wrapper = shallow(<Block {...props} />);
    expect(wrapper.find(AuthorList)).toHaveLength(1);
  });
});