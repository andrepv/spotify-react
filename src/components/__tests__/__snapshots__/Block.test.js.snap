// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Block /> renders correctly 1`] = `
<div
  className="block"
>
  <div
    className="block__media"
  >
    <div
      className="block__img block__img_root bg-center"
      style={
        Object {
          "backgroundImage": "url(test.png)",
        }
      }
    />
  </div>
  <div
    className="block__content"
  >
    <a
      href="/playlist/1"
      onClick={[Function]}
    >
      <span
        className="block__title"
        onClick={[MockFunction]}
      >
        test
      </span>
    </a>
    <p
      className="block__subtitle"
    >
      0 tracks
    </p>
  </div>
</div>
`;
